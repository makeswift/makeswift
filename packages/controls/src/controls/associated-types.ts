export type AssociatedTypes<T> = T extends {
  __associatedTypes(types: infer Types): void
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]

export type ControlType<D> = AssociatedType<D, 'ControlType'>
export type ConfigType<D> = AssociatedType<D, 'Config'>
export type DataType<D> = AssociatedType<D, 'DataType'>
export type ValueType<D> = AssociatedType<D, 'ValueType'>
export type ResolvedValueType<D> = AssociatedType<D, 'ResolvedValueType'>
export type InstanceType<D> = AssociatedType<D, 'InstanceType'>
