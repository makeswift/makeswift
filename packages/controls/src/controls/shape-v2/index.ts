export * from './shape-v2'
export * from './shape-v2-control'
