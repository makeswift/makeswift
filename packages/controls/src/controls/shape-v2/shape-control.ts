import { ControlDefinition } from '../definition'
import {
  ControlInstance,
  type ControlMessage,
  type SendMessage,
} from '../instance'

import { ShapeDefinition } from './shape'

type Message = {
  type: typeof ShapeControl.CHILD_CONTROL_MESSAGE
  payload: { message: ControlMessage; key: string }
}

export class ShapeControl<
  Def extends ShapeDefinition = ShapeDefinition,
> extends ControlInstance<Message> {
  static readonly CHILD_CONTROL_MESSAGE =
    'makeswift::controls::shape::message::child-control-message'

  private readonly childControls: Map<string, ControlInstance> = new Map()

  constructor(
    private readonly definition: Def,
    sendMessage: SendMessage<Message>,
  ) {
    super(sendMessage)
    this.childControls = new Map(
      Object.entries(this.definition.keyDefs).map(([key, def]) => [
        key,
        this.createChildControl(def, key),
      ]),
    )
  }

  recv = (message: Message) => {
    switch (message.type) {
      case ShapeControl.CHILD_CONTROL_MESSAGE: {
        this.child(message.payload.key)?.recv(message.payload.message)
      }
    }
  }

  child = (key: string) => this.childControls.get(key)

  createChildControl = (def: ControlDefinition, key: string) => {
    return def.createInstance((message) =>
      this.sendMessage({
        type: ShapeControl.CHILD_CONTROL_MESSAGE,
        payload: { message, key },
      }),
    )
  }
}
