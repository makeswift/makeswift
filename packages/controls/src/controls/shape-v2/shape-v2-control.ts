import { ControlDefinition } from '../definition'
import {
  ControlInstance,
  type ControlMessage,
  type SendMessage,
} from '../instance'

import { ShapeV2Definition } from './shape-v2'

type Message = {
  type: typeof ShapeV2Control.CHILD_CONTROL_MESSAGE
  payload: { message: ControlMessage; key: string }
}

export class ShapeV2Control<
  Def extends ShapeV2Definition = ShapeV2Definition,
> extends ControlInstance<Message> {
  static readonly CHILD_CONTROL_MESSAGE =
    'makeswift::controls::shape-v2::message::child-control-message'

  private readonly childControls: Map<string, ControlInstance> = new Map()

  constructor(
    private readonly definition: Def,
    sendMessage: SendMessage<Message>,
  ) {
    super(sendMessage)
    this.childControls = new Map(
      Object.entries(this.definition.keyDefs).map(([key, def]) => [
        key,
        this.createChildControl(def, key),
      ]),
    )
  }

  recv = (message: Message) => {
    switch (message.type) {
      case ShapeV2Control.CHILD_CONTROL_MESSAGE: {
        this.child(message.payload.key)?.recv(message.payload.message)
      }
    }
  }

  child = (key: string) => this.childControls.get(key)

  createChildControl = (def: ControlDefinition, key: string) => {
    return def.createInstance((message) =>
      this.sendMessage({
        type: ShapeV2Control.CHILD_CONTROL_MESSAGE,
        payload: { message, key },
      }),
    )
  }
}
