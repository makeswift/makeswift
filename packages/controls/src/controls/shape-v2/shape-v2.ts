import { z } from 'zod'

import { mapValues } from '../../lib/functional'
import { StableValue } from '../../lib/stable-value'
import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey, type Data } from '../../common'
import {
  type CopyContext,
  type MergeTranslatableDataContext,
} from '../../context'
import { type IntrospectionTarget } from '../../introspection'
import { type ResourceResolver } from '../../resources/resolver'
import {
  SerializationSchema,
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'
import { type Stylesheet } from '../../stylesheet'

import {
  type DataType as DataType_,
  type ResolvedValueType as ResolvedValueType_,
  type ValueType as ValueType_,
} from '../associated-types'
import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { type SendMessage } from '../instance'

import { ShapeV2Control } from './shape-v2-control'

type ItemDefinition = ControlDefinition<string, unknown, any, any, any>
type KeyDefinitions = Record<string, ItemDefinition>

export type Config<Defs extends KeyDefinitions = KeyDefinitions> = {
  label?: string
  layout?: typeof Definition.Layout.Inline | typeof Definition.Layout.Popover
  readonly type: Defs
}

export type ShapeV2DataTypeV0<C extends Config> = {
  [K in keyof C['type']]?: DataType_<C['type'][K]>
}

export const ShapeV2DataTypeV1VersionId = 'shape-v2::v1' as const

export type ShapeV2DataTypeV1<C extends Config> = {
  [ControlDataTypeKey]: typeof ShapeV2DataTypeV1VersionId
  value: {
    [K in keyof C['type']]?: DataType_<C['type'][K]>
  }
}

type DataType<C extends Config> = ShapeV2DataTypeV0<C> | ShapeV2DataTypeV1<C>

type ValueType<C extends Config> = {
  [K in keyof C['type']]?: ValueType_<C['type'][K]>
}

type ResolvedValueType<C extends Config> = {
  [K in keyof C['type']]: ResolvedValueType_<C['type'][K]>
}

type InstanceType<C extends Config> = ShapeV2Control<Definition<C>>

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>,
  InstanceType<C>
> {
  private static readonly v1DataType = ShapeV2DataTypeV1VersionId
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const
  static readonly type = 'makeswift::controls::shape-v2' as const

  static readonly Layout = {
    Inline: `${this.type}::layout::inline`,
    Popover: `${this.type}::layout::popover`,
  } as const

  static deserialize(
    data: DeserializedRecord,
    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
  ): ShapeV2Definition {
    if (data.type !== Definition.type)
      throw new Error(
        `Shape: expected '${Definition.type}', got '${data.type}'`,
      )

    const {
      config: { type, ...config },
    } = z
      .object({
        type: z.literal(Definition.type),
        config: z.object({
          label: z.string().optional(),
          layout: z
            .union([
              z.literal(Definition.Layout.Inline),
              z.literal(Definition.Layout.Popover),
            ])
            .optional(),
          type: z.record(z.string(), SerializationSchema.deserializedRecord),
        }),
      })
      .parse(data)

    const deserializedType = mapValues(type, (itemDef) => {
      return deserializeCallback(itemDef)
    })

    return ShapeV2({
      type: deserializedType,
      ...config,
    })
  }

  get keyDefs() {
    return this.config.type
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    const version = z.literal(1)
    const type = z.literal(Definition.type)

    const keys = mapValues(this.keyDefs, (def) => def.schema)

    const v0Data = z.object(mapValues(keys, (key) => key.data)) as SchemaType<
      ShapeV2DataTypeV0<C>
    >
    const v1Data = z.object({
      [ControlDataTypeKey]: z.literal(Definition.v1DataType),
      value: v0Data,
    }) as SchemaType<ShapeV2DataTypeV1<C>>

    const data = z.union([v0Data, v1Data])

    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
      ValueType<C>
    >

    const resolvedValue = z.object(
      mapValues(keys, (key) => key.resolvedValue),
    ) as SchemaType<ResolvedValueType<C>>

    const config = z.object({
      label: z.string().optional(),
      layout: z.union([
        z.literal(Definition.Layout.Inline),
        z.literal(Definition.Layout.Popover),
      ]),
      type: z.object(mapValues(keys, (key) => key.definition)),
    })

    const definition = z.object({
      type,
      config,
    })

    return { type, data, value, resolvedValue, definition, version }
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    if (data == null) return undefined

    if (
      ControlDataTypeKey in data &&
      data[ControlDataTypeKey] === Definition.v1DataType
    ) {
      return mapValues(this.keyDefs, (def, key) =>
        def.fromData((data as ShapeV2DataTypeV1<C>).value[key]),
      )
    }

    return mapValues(this.keyDefs, (def, key) =>
      def.fromData((data as ShapeV2DataTypeV0<C>)[key]),
    )
  }

  toData(value: ValueType<C>): DataType<C> {
    return {
      ...Definition.dataSignature.v1,
      value: mapValues(this.keyDefs, (def, key) => def.toData(value[key])),
    }
  }

  copyData(
    data: DataType<C> | undefined,
    context: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return undefined

    if (
      ControlDataTypeKey in data &&
      data[ControlDataTypeKey] === Definition.v1DataType
    ) {
      return mapValues(this.keyDefs, (def, key) =>
        def.copyData((data as ShapeV2DataTypeV1<C>).value[key], context),
      )
    }

    return mapValues(this.keyDefs, (def, key) =>
      def.copyData((data as ShapeV2DataTypeV0<C>)[key], context),
    )
  }

  getTranslatableData(data: DataType<C>): Data {
    if (data == null) return null

    if (
      ControlDataTypeKey in data &&
      data[ControlDataTypeKey] === Definition.v1DataType
    ) {
      return mapValues(this.keyDefs, (def, key) =>
        def.getTranslatableData((data as ShapeV2DataTypeV1<C>).value[key]),
      )
    }

    return mapValues(this.keyDefs, (def, key) =>
      def.getTranslatableData((data as ShapeV2DataTypeV0<C>)[key]),
    )
  }

  mergeTranslatedData(
    data: DataType<C>,
    translatedData: Record<string, DataType<C>>,
    context: MergeTranslatableDataContext,
  ): Data {
    if (translatedData == null) return data

    if (
      ControlDataTypeKey in data &&
      data[ControlDataTypeKey] === Definition.v1DataType
    ) {
      return mapValues(this.keyDefs, (def, key) =>
        def.mergeTranslatedData(
          (data as ShapeV2DataTypeV1<C>).value[key],
          translatedData[key],
          context,
        ),
      )
    }

    return mapValues(this.keyDefs, (def, key) =>
      def.mergeTranslatedData(
        (data as ShapeV2DataTypeV0<C>)[key],
        translatedData[key],
        context,
      ),
    )
  }

  resolveValue(
    data: DataType<C> | undefined,
    resolver: ResourceResolver,
    stylesheet: Stylesheet,
    control?: InstanceType<C>,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    const propsData = data != null ? Definition.propsData(data) : data

    const keyValues = mapValues(this.keyDefs, (def, key) =>
      def.resolveValue(
        propsData?.[key],
        resolver,
        stylesheet.child(key),
        control?.child(key),
      ),
    )

    const stableValue = StableValue({
      name: Definition.type,
      read: () =>
        mapValues(keyValues, (v) => v.readStable()) as ResolvedValueType<C>,
      deps: Object.values(keyValues),
    })

    return {
      ...stableValue,
      triggerResolve: async (currentValue?: ResolvedValueType<C>) => {
        await Promise.all(
          Object.entries(keyValues).map(([key, v]) =>
            v?.triggerResolve(currentValue?.[key]),
          ),
        )
      },
    }
  }

  createInstance(sendMessage: SendMessage): InstanceType<C> {
    return new ShapeV2Control(this, sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  introspect<R>(data: DataType<C> | undefined, target: IntrospectionTarget<R>) {
    if (data == null) return []

    if (
      ControlDataTypeKey in data &&
      data[ControlDataTypeKey] === Definition.v1DataType
    ) {
      return Object.entries(this.keyDefs).flatMap(
        ([key, def]) =>
          def.introspect((data as ShapeV2DataTypeV1<C>).value[key], target) ??
          [],
      )
    }

    return Object.entries(this.keyDefs).flatMap(
      ([key, def]) =>
        def.introspect((data as ShapeV2DataTypeV0<C>)[key], target) ?? [],
    )
  }

  static propsData<C extends Config>(data: DataType<C>): ShapeV2DataTypeV0<C> {
    if (Definition.isVersionedData(data)) {
      return data.value
    }

    return data
  }

  private static isVersionedData<C extends Config>(
    data: DataType<C>,
  ): data is ShapeV2DataTypeV1<C> {
    return (
      data != null &&
      ControlDataTypeKey in data &&
      data[ControlDataTypeKey] === Definition.v1DataType
    )
  }
}

type UserConfig<
  D extends KeyDefinitions,
  L extends Config['layout'],
> = Config<D> & {
  layout?: L
}

type NormedConfig<
  D extends KeyDefinitions,
  L extends Config['layout'] = Config['layout'],
> = undefined extends L
  ? Config<D>
  : Pick<Config<D>, 'label' | 'type'> & {
      layout: L
    }

export class ShapeV2Definition<
  C extends Config = Config,
> extends Definition<C> {}

export function ShapeV2<D extends KeyDefinitions, L extends Config['layout']>(
  config: UserConfig<D, L>,
): ShapeV2Definition<NormedConfig<D, L>> {
  return new ShapeV2Definition<NormedConfig<D, L>>(
    (config?.layout == null
      ? { layout: Definition.Layout.Inline, ...config }
      : config) as NormedConfig<D, L>,
  )
}

ShapeV2.Layout = Definition.Layout
