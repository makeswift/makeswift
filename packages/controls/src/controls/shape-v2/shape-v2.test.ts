import { ControlDataTypeKey } from '../../common'
import { createReplacementContext } from '../../context'
import { Targets } from '../../introspection'
import { deserializeRecord, type DeserializedRecord } from '../../serialization'

import { type DataType, type ValueType } from '../associated-types'
import { Checkbox, CheckboxDefinition } from '../checkbox'
import { Color, ColorDefinition } from '../color'
import { Combobox } from '../combobox'
import { ControlDefinition } from '../definition'
import { Image } from '../image'
import { GenericLink as Link } from '../link'
import { List } from '../list'
import { Number } from '../number'
import { TextArea } from '../text-area'
import { TextInput } from '../text-input'

import { ShapeV2, ShapeV2Definition } from './shape-v2'

function deserializer(record: DeserializedRecord): ControlDefinition {
  switch (record.type) {
    case ShapeV2Definition.type:
      return ShapeV2Definition.deserialize(record, deserializer)

    case ColorDefinition.type:
      return ColorDefinition.deserialize(record)

    case CheckboxDefinition.type:
      return CheckboxDefinition.deserialize(record)

    default:
      throw new Error('Invalid control type')
  }
}

describe('ShapeV2', () => {
  describe('constructor', () => {
    test('heterogenous controls', () => {
      const shape = ShapeV2({
        type: {
          val: Color({ defaultValue: 'red' }),
          num: Checkbox(),
        },
      })
      expect(shape).toMatchSnapshot()
    })

    test('disallows extraneous properties', () => {
      ShapeV2({
        type: {
          toggle: Checkbox(),
        },
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: ShapeV2Definition) {}
    assignTest(
      ShapeV2({
        type: {},
        label: 'Shape',
      }),
    )

    assignTest(
      ShapeV2({
        type: {},
      }),
    )

    assignTest(
      ShapeV2({
        type: {
          toggle: Checkbox(),
          color: Color({ defaultValue: 'red' }),
        },
      }),
    )

    assignTest(
      ShapeV2({
        type: {
          combo: Combobox({
            getOptions: () => [{ id: '0', value: 'a', label: 'Alpha' }],
          }),
          link: Link({ label: 'Link' }),
        },
      }),
    )
  })

  describe('serialization', () => {
    test('serialize/deserialize layout of controls', () => {
      const shape = ShapeV2({
        type: {},
        layout: 'makeswift::controls::shape-v2::layout::popover',
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeV2Definition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })

    test('serialize/deserialize label of controls', () => {
      const shape = ShapeV2({
        type: {},
        label: 'Shape',
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeV2Definition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })

    test('serialize/deserialize shape of controls', () => {
      const shape = ShapeV2({
        type: {
          color: Color({ defaultValue: 'red' }),
          checkbox: Checkbox(),
        },
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeV2Definition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })

    test('serialize/deserialize composable shape of controls', () => {
      const shape = ShapeV2({
        type: {
          shape: ShapeV2({
            type: {
              color: Color({ defaultValue: 'red' }),
              checkbox: Checkbox(),
            },
          }),
        },
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeV2Definition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })
  })

  describe('introspection', () => {
    test('Shape data is correctly delegated to subcontrols', () => {
      const shape = ShapeV2({
        type: {
          color: Color({ defaultValue: 'red' }),
          link: Link(),
          image: Image(),
        },
      })

      const shapeData = {
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
        link: {
          type: 'OPEN_PAGE' as const,
          payload: {
            pageId: 'page-id',
            openInNewTab: false,
          },
        },
        image: 'file-id',
      }

      const swatchIds = shape.introspect(shapeData, Targets.Swatch)
      expect(swatchIds).toEqual(['swatch-id'])

      const pageIds = shape.introspect(shapeData, Targets.Page)
      expect(pageIds).toEqual(['page-id'])

      const fileIds = shape.introspect(shapeData, Targets.File)
      expect(fileIds).toEqual(['file-id'])
    })

    test('ShapeV2 versioned data is correctly delegated to subcontrols', () => {
      const shape = ShapeV2({
        type: {
          color: Color({ defaultValue: 'red' }),
          link: Link(),
          image: Image(),
        },
      })

      const shapeData = {
        [ControlDataTypeKey]: 'shape-v2::v1' as const,
        value: {
          color: {
            swatchId: 'swatch-id',
            alpha: 1,
          },
          link: {
            type: 'OPEN_PAGE' as const,
            payload: {
              pageId: 'page-id',
              openInNewTab: false,
            },
          },
          image: 'file-id',
        },
      }

      const swatchIds = shape.introspect(shapeData, Targets.Swatch)
      expect(swatchIds).toEqual(['swatch-id'])

      const pageIds = shape.introspect(shapeData, Targets.Page)
      expect(pageIds).toEqual(['page-id'])

      const fileIds = shape.introspect(shapeData, Targets.File)
      expect(fileIds).toEqual(['file-id'])
    })
  })

  describe('getTranslatableData', () => {
    test('Shape data is correctly delegated to subcontrols', () => {
      const shape = ShapeV2({
        type: {
          textArea: TextArea(),
          textInput: TextInput(),
          color: Color({ defaultValue: 'red' }),
        },
      })

      const shapeData = {
        textArea: 'alpha',
        textInput: 'beta',
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
      }

      const translatableData = shape.getTranslatableData(shapeData)
      expect(translatableData).toEqual({
        textArea: 'alpha',
        textInput: 'beta',
        color: null,
      })
    })

    test('ShapeV2 versioned data is correctly delegated to subcontrols', () => {
      const shape = ShapeV2({
        type: {
          textArea: TextArea(),
          textInput: TextInput(),
          color: Color({ defaultValue: 'red' }),
        },
      })

      const shapeData = {
        [ControlDataTypeKey]: 'shape-v2::v1' as const,
        value: {
          textArea: 'alpha',
          textInput: 'beta',
          color: {
            swatchId: 'swatch-id',
            alpha: 1,
          },
        },
      }

      const translatableData = shape.getTranslatableData(shapeData)
      expect(translatableData).toEqual({
        textArea: 'alpha',
        textInput: 'beta',
        color: null,
      })
    })
  })

  describe('Shape data is gracefully handles missing/extra props', () => {
    const shape = ShapeV2({
      type: {
        color: Color({ defaultValue: 'red' }),
        list: List({ type: Number() }),
        link: Link(),
        text: TextArea(),
      },
    })

    const data = {
      // missing props + extra prop that should be ignored
      color: { swatchId: '[swatch-id]', alpha: 1 },
      text: 'Hello world',
      extraProp: 'extra',
    } as DataType<typeof shape>

    const value = {
      color: { swatchId: '[swatch-id]', alpha: 1 },
      list: [17],
      link: {
        type: 'OPEN_PAGE',
        payload: { pageId: '[page-id]', openInNewTab: false },
      },
      text: 'Hello world',
      // extra prop only, `toData` doesn't handle missing props
      extraProp: 'extra',
    } as ValueType<typeof shape>

    test('fromData', () => expect(shape.fromData(data)).toMatchSnapshot())
    test('toData', () => expect(shape.toData(value)).toMatchSnapshot())
    test('copyData', () =>
      expect(
        shape.copyData(data, {
          replacementContext: createReplacementContext({}),
          copyElement: (node) => node,
        }),
      ).toMatchSnapshot())

    test('getTranslatableData', () =>
      expect(shape.getTranslatableData(data)).toMatchSnapshot())

    test('mergeTranslatedData', () =>
      expect(
        shape.mergeTranslatedData(
          data,
          {},
          {
            translatedData: {},
            mergeTranslatedData: (node) => node,
          },
        ),
      ).toMatchSnapshot())

    test('introspect', () =>
      expect(shape.introspect(data, Targets.ChildrenElement)).toStrictEqual([]))
  })

  describe('ShapeV2 versioned data is gracefully handles missing/extra props', () => {
    const shape = ShapeV2({
      type: {
        color: Color({ defaultValue: 'red' }),
        list: List({ type: Number() }),
        link: Link(),
        text: TextArea(),
      },
    })

    const data = {
      [ControlDataTypeKey]: 'shape-v2::v1' as const,
      // missing props + extra prop that should be ignored
      value: {
        color: { swatchId: '[swatch-id]', alpha: 1 },
        text: 'Hello world',
        extraProp: 'extra',
      },
    } as DataType<typeof shape>

    const value = {
      color: { swatchId: '[swatch-id]', alpha: 1 },
      list: [17],
      link: {
        type: 'OPEN_PAGE',
        payload: { pageId: '[page-id]', openInNewTab: false },
      },
      text: 'Hello world',
      // extra prop only, `toData` doesn't handle missing props
      extraProp: 'extra',
    } as ValueType<typeof shape>

    test('fromData', () => expect(shape.fromData(data)).toMatchSnapshot())
    test('toData', () => expect(shape.toData(value)).toMatchSnapshot())
    test('copyData', () =>
      expect(
        shape.copyData(data, {
          replacementContext: createReplacementContext({}),
          copyElement: (node) => node,
        }),
      ).toMatchSnapshot())

    test('getTranslatableData', () =>
      expect(shape.getTranslatableData(data)).toMatchSnapshot())

    test('mergeTranslatedData', () =>
      expect(
        shape.mergeTranslatedData(
          data,
          {},
          {
            translatedData: {},
            mergeTranslatedData: (node) => node,
          },
        ),
      ).toMatchSnapshot())

    test('introspect', () =>
      expect(shape.introspect(data, Targets.ChildrenElement)).toStrictEqual([]))
  })
})
