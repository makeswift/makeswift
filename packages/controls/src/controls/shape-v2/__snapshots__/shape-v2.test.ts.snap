// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShapeV2 Shape data is gracefully handles missing/extra props copyData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`ShapeV2 Shape data is gracefully handles missing/extra props fromData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`ShapeV2 Shape data is gracefully handles missing/extra props getTranslatableData 1`] = `
{
  "color": null,
  "link": null,
  "list": null,
  "text": "Hello world",
}
`;

exports[`ShapeV2 Shape data is gracefully handles missing/extra props mergeTranslatedData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`ShapeV2 Shape data is gracefully handles missing/extra props toData 1`] = `
{
  "@@makeswift/type": "shape-v2::v1",
  "value": {
    "color": {
      "@@makeswift/type": "color::v1",
      "alpha": 1,
      "swatchId": "[swatch-id]",
    },
    "link": {
      "payload": {
        "openInNewTab": false,
        "pageId": "[page-id]",
      },
      "type": "OPEN_PAGE",
    },
    "list": [
      {
        "id": "xxxxxxxx-xxxx-xxxx-xxxx-100000000000",
        "type": "makeswift::controls::number",
        "value": {
          "@@makeswift/type": "number::v1",
          "value": 17,
        },
      },
    ],
    "text": {
      "@@makeswift/type": "text-area::v1",
      "value": "Hello world",
    },
  },
}
`;

exports[`ShapeV2 ShapeV2 versioned data is gracefully handles missing/extra props copyData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`ShapeV2 ShapeV2 versioned data is gracefully handles missing/extra props fromData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`ShapeV2 ShapeV2 versioned data is gracefully handles missing/extra props getTranslatableData 1`] = `
{
  "color": null,
  "link": null,
  "list": null,
  "text": "Hello world",
}
`;

exports[`ShapeV2 ShapeV2 versioned data is gracefully handles missing/extra props mergeTranslatedData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`ShapeV2 ShapeV2 versioned data is gracefully handles missing/extra props toData 1`] = `
{
  "@@makeswift/type": "shape-v2::v1",
  "value": {
    "color": {
      "@@makeswift/type": "color::v1",
      "alpha": 1,
      "swatchId": "[swatch-id]",
    },
    "link": {
      "payload": {
        "openInNewTab": false,
        "pageId": "[page-id]",
      },
      "type": "OPEN_PAGE",
    },
    "list": [
      {
        "id": "xxxxxxxx-xxxx-xxxx-xxxx-100000000001",
        "type": "makeswift::controls::number",
        "value": {
          "@@makeswift/type": "number::v1",
          "value": 17,
        },
      },
    ],
    "text": {
      "@@makeswift/type": "text-area::v1",
      "value": "Hello world",
    },
  },
}
`;

exports[`ShapeV2 constructor heterogenous controls 1`] = `
ShapeV2Definition {
  "config": {
    "layout": "makeswift::controls::shape-v2::layout::inline",
    "type": {
      "num": CheckboxDefinition {
        "config": {},
        "version": 1,
      },
      "val": ColorDefinition {
        "config": {
          "defaultValue": "red",
        },
        "version": 1,
      },
    },
  },
}
`;

exports[`ShapeV2 serialization serialize/deserialize composable shape of controls 1`] = `
{
  "config": {
    "layout": "makeswift::controls::shape-v2::layout::inline",
    "type": {
      "shape": {
        "config": {
          "layout": "makeswift::controls::shape-v2::layout::inline",
          "type": {
            "checkbox": {
              "config": {},
              "type": "makeswift::controls::checkbox",
              "version": 1,
            },
            "color": {
              "config": {
                "defaultValue": "red",
              },
              "type": "makeswift::controls::color",
              "version": 1,
            },
          },
        },
        "type": "makeswift::controls::shape-v2",
      },
    },
  },
  "type": "makeswift::controls::shape-v2",
}
`;

exports[`ShapeV2 serialization serialize/deserialize label of controls 1`] = `
{
  "config": {
    "label": "Shape",
    "layout": "makeswift::controls::shape-v2::layout::inline",
    "type": {},
  },
  "type": "makeswift::controls::shape-v2",
}
`;

exports[`ShapeV2 serialization serialize/deserialize layout of controls 1`] = `
{
  "config": {
    "layout": "makeswift::controls::shape-v2::layout::popover",
    "type": {},
  },
  "type": "makeswift::controls::shape-v2",
}
`;

exports[`ShapeV2 serialization serialize/deserialize shape of controls 1`] = `
{
  "config": {
    "layout": "makeswift::controls::shape-v2::layout::inline",
    "type": {
      "checkbox": {
        "config": {},
        "type": "makeswift::controls::checkbox",
        "version": 1,
      },
      "color": {
        "config": {
          "defaultValue": "red",
        },
        "type": "makeswift::controls::color",
        "version": 1,
      },
    },
  },
  "type": "makeswift::controls::shape-v2",
}
`;
